### PR TITLE
feat: demand fill tracking

### PR DESCRIPTION
**Pull request by kosecki123**
This PR provides the methods to calculate missing energy in demand in total and per period. `market-matcher` is now skipping demand with the current period already filled. 

The solution is based on `DemandPartiallyFilled` event from the `market`, uses it to aggregate information about filled demand periods. Event emit timestamp is assumed to be the time of fill operation.

This also fixes the bug where only certificates with higher or matching energy were matched with the demand.

Untitled Diagram

### CODE REVIEW

File: packages/market/src/blockchain-facade/Demand.ts [hunk 1]
```diff
@@ -229,3 +250,108 @@ export const deleteDemand = async (
...
+};
+
+export const calculateMissingEnergyDemand = async (
+    demand: IDemand,
+    config: Configuration.Entity
+) => {
+    const { startTime, endTime, timeFrame, energyPerTimeFrame } = demand.offChainProperties;
+
+    if (timeFrame === TimeFrame.hourly || timeFrame === TimeFrame.halfHourly) {
+        throw new Error('Hourly and half-hourly demands are not supported');
```
Comment (josipbagaric) [inline packages/market/src/blockchain-facade/Demand.ts]: Why not?

**Review by josipbagaric — APPROVED** [review]

File: packages/asset-registry/src/blockchain-facade/ConsumingAsset.ts [hunk 2]
```diff
@@ -72,7 +72,7 @@ export const getAllAssetsOwnedBy = async (owner: string, configuration: Configur
 
 export class Entity extends Asset.Entity implements Asset.IOnChainProperties {
     getUrl(): string {
-        const consumingAssetLogicAddress = this.configuration.blockchainProperties.consumingAssetLogicInstance.web3Contract._address;
+        const consumingAssetLogicAddress = this.configuration.blockchainProperties.consumingAssetLogicInstance.web3Contract.options.address;
```
Comment (e00dan) [inline packages/asset-registry/src/blockchain-facade/ConsumingAsset.ts:75]: Is this preferred way of accessing address?

File: packages/event-listener/src/listeners/origin.listener.ts [hunk 3]
```diff
@@ -204,8 +204,8 @@ export class OriginEventListener implements IOriginEventListener {
             this.conf
         ).sync();
 
-        const matchedDemands = demands.filter(demand => {
-            const { result } = new MatchableDemand(demand).matchesCertificate(
+        const matchedDemands = demands.filter(async demand => {
+            const { result } = await new MatchableDemand(demand).matchesCertificate(
```
Comment (e00dan) [inline packages/event-listener/src/listeners/origin.listener.ts]: Is this construct going to work? `await new` looks unfamiliar to me

File: packages/origin-ui/src/__tests__/DemandForm.test.tsx [hunk 4]
```diff
@@ -210,13 +218,13 @@ describe('DemandForm', () => {
         expect(submitButton.getDOMNode().hasAttribute('disabled')).toBe(false);
 
         createDemand = (offChainProps: Demand.IDemandOffChainProperties) => {
-            expect(moment.unix(parseInt(offChainProps.startTime, 10)).format('YYYY-MM-D')).toBe(
+            expect(moment.unix(offChainProps.startTime).format('YYYY-MM-D')).toBe(
```
Comment (e00dan) [inline packages/origin-ui/src/__tests__/DemandForm.test.tsx:221]: :clap:

File: packages/origin-ui/src/components/Demand/DemandForm.tsx [hunk 5]
```diff
@@ -41,34 +41,6 @@ import { FormikDatePicker } from '../FormikDatePicker';
 import { AssetTypeSelector } from '../AssetTypeSelector';
 import { getCurrentUser } from '../../features/users/selectors';
 
-export function calculateTotalEnergyDemand(
```
Comment (e00dan) [inline packages/origin-ui/src/components/Demand/DemandForm.tsx:44]: :clap:

**Review by e00dan — APPROVED** [review]
:+1:

File: packages/asset-registry/src/blockchain-facade/ConsumingAsset.ts (hunk 2 above)
Comment (kosecki123) [reply packages/asset-registry/src/blockchain-facade/ConsumingAsset.ts:75]: Yeah. That seems to be the only way if you follow the `Contract` type. which we now do.

File: packages/event-listener/src/listeners/origin.listener.ts (hunk 3 above)
Comment (kosecki123) [reply packages/event-listener/src/listeners/origin.listener.ts]: `new` has higher precedence than `await` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence#Table

**Pull request merged**